src/StackExchange.Redis/ConfigurationOptions.cs

```diff
@@ -296,15 +296,26 @@ private static RemoteCertificateValidationCallback TrustIssuerCallback(X509Certi
         {
             if (issuer == null) throw new ArgumentNullException(nameof(issuer));
 
-            return (object _, X509Certificate? certificate, X509Chain? __, SslPolicyErrors sslPolicyError)
-                => sslPolicyError == SslPolicyErrors.RemoteCertificateChainErrors
-                    && certificate is X509Certificate2 v2
-                    && CheckTrustedIssuer(v2, issuer);
+            return (object _, X509Certificate? certificate, X509Chain? __, SslPolicyErrors sslPolicyError) =>
+            {
+                // If we're already valid, there's nothing further to check
+                if (sslPolicyError == SslPolicyErrors.None)
+                {
+                    return true;
+                }
+                // If we're not valid due to chain errors - check against the trusted issuer
+                // Note that we're only proceeding at all here if the *only* issue is chain errors (not more flags in SslPolicyErrors)
+                return sslPolicyError == SslPolicyErrors.RemoteCertificateChainErrors
+                       && certificate is X509Certificate2 v2
+                       && CheckTrustedIssuer(v2, issuer);
+            };
         }
 
         private static bool CheckTrustedIssuer(X509Certificate2 certificateToValidate, X509Certificate2 authority)
```

Comment:
OK.  Last comment (I think :wink:).

Aside from Internet connectivity flakiness, the only difference between this chain build and the one you were given in the callback is that you've definitely added the specified pin/anchor to ExtraStore.  That means the chain will realistically only be different if the incoming chain terminated in PartialChain and the specified pin/anchor is the next hop.

_That_ means that you could run this same logic on the existing chain, with two minor changes:
1) You'd need to check each chainElement.ChainElementStatus for the presence of an error other than `PartialChain` or `UnknownRootAuthority`.
2) If it did have PartialChain, and you didn't find your anchor, _then_ you'd run it again with this logic.

Both the flow I suggested here and what the code is trying to accomplish, you'll potentially have issues regarding cross-certification.  Since you can't ask .NET to ask the OS to consider this intermediate as a trust anchor, the OS might prefer a variant of the certificate.  For example, the Let's Encrypt R3 intermediate was signed by both "ISRG Root X1" and "DST Root CA X3" (https://crt.sh/?caid=183267).  If the anchor in your flow was the R3 signed by DST-X3; but the underlying OS trusts ISRG-X1 and not DST-X3, then the OS will completely ignore the ExtraStore input as irrelevant because it always prefers to build trusted chains over untrusted chains.  If it trusts both, or neither, then "stuff happens" (I don't know if the ExtraStore nudge will act as a tiebreaker, or if it comes down to load-order, phase of the moon, etc).  Really this is just a concern for the inputs: the certificates should always be from nice single-path segments of a chain.  Once you've made that constraint/assumption, then you don't also have a reason to rerun the chain build just because you don't see the intended anchor.

And, finally, regarding disposing things... for maximum reduction of finalization, you should dispose all of the certificates from chain.ChainElements in the chain the callback gave you.  .NET Framework didn't dispose them (so they went off to the finalizer).  And I thought the .NET Core line only did it when you didn't have a callback; but looking at the code just now it seems to do it in a finally and has no suppression for if the chain was passed to a callback.  Huh.